File: tests/dovydas.xyz/objects/pages/AbstractPage.ts
import { Page, Locator, expect } from '@playwright/test'


export abstract class AbstractPage {
  readonly page: Page
  readonly pageName: string
  readonly pagePath: string

  constructor(page: Page, pageName: string, pagePath: string) {
    this.page = page
    this.pageName = pageName
    this.pagePath = pagePath
  }

  async wait(time: number) {
    await this.page.waitForTimeout(time)
  }

  getSnapshotName(): string {
    return `${this.pageName}.png`
  }

  async visit() {
    await this.page.goto(this.pagePath)
  }

  throwOnMissingCase(locatorConstant: any){
    throw new Error(`Non-existent locator: ${locatorConstant}`)
  }
  

  abstract selectLocator(locatorConstant: any): Locator

  async clickOnLink(locatorConstant: any){
    await this.selectLocator(locatorConstant).click()
  }

  async assertAttributeValue(locatorConstant: any, attribute: string, valueExpected: string) {
    await expect(this.selectLocator(locatorConstant)).toHaveAttribute(attribute, valueExpected)
  }

  async assertLinkText(locatorConstant: any, valueExpected: string){
    await expect(this.selectLocator(locatorConstant)).toHaveText(valueExpected)
  }

  async assertRegexPatternMatches(locatorConstant: any, regex: RegExp){
    const text = await this.selectLocator(locatorConstant).textContent()
    expect(text).toEqual(expect.stringMatching(regex))
  }

  async assertVisible(locatorConstant: any){
    await expect(this.selectLocator(locatorConstant)).toBeVisible()
  }


}

File: dovydas.xyz/tests/objects/abstract_page.py
import re

from abc import abstractmethod
from enum import Enum
from typing import Any, Pattern

from playwright.async_api import Locator, Page, expect


class AbstractPage:
    def __init__(self, page: Page) -> None:
        self.page = page

    @property
    @abstractmethod
    def page_name(self) -> str: ...

    @property
    @abstractmethod
    def page_path(self) -> str: ...

    @abstractmethod
    def select_locator(self, locator_constant: Any) -> Locator: ...

    async def wait(self, time: int) -> None:
        await self.page.wait_for_timeout(time)

    def get_snapshot_name(self) -> str:
        return f"{self.page_name}.png"

    async def visit(self) -> None:
        await self.page.goto(self.page_path)

    def throw_on_missing_case(self, locator_constant: Any) -> None:
        raise ValueError(f"Non-existent locator: {locator_constant}")

    async def click_on_link(self, locator_constant: Any) -> None:
        await self.select_locator(locator_constant).click()

    async def assert_attribute_value(self, locator_constant: Any, attribute: str, value_expected: str) -> None:
        await expect(self.select_locator(locator_constant)).to_have_attribute(attribute, value_expected)

    async def assert_link_text(self, locator_constant: Any, value_expected: str) -> None:
        await expect(self.select_locator(locator_constant)).to_have_text(value_expected)

    async def assert_regex_pattern_matches(self, locator_constant: Any, regex: Pattern[str]) -> None:
        text = await self.select_locator(locator_constant).text_content()
        assert text
        assert re.search(regex, text), f"Text '{text}' does not match regex '{regex}'"

    async def assert_visible(self, locator_constant: Any) -> None:
        await expect(self.select_locator(locator_constant)).to_be_visible()
